### ISSUE TITLE
(Sub)Domain-based Routing

### ISSUE THREAD

**Issue by hrueger** [sveltejs/kit#8335]
### Describe the problem

Hi,
first of all, thanks for this excellent piece of software! I truely love it.

I'm currently building a platform where I have multiple tenants. They can configure some settings and provide a frontend for their users. This should live on `https://tenant-slug.my-platform.tld`. Then, they should also have an admin-dashboard at `https://admin.tenant-slug.my-platform.tld`.
I'll also have some registration / onboarding ui at `https://registration.my-platform.tld`.

Currently, this does not seem possible with a single Svelte app. I can't build one for each tenant, because they change quite often. 

At DNS level it would be solved via a wildcard subdomain and a wildcard SSL certificate.

### Describe the proposed solution

As proxies (see below) are complicated to setup and didn't work for my usecase, I'd like to be able to have different routes for each subdomain and the pure domain.
Maybe this is kind-of linked: https://vercel.com/guides/nextjs-multi-tenant-application

**Here are my thoughts:**
Currently, there's the `routes` directory with all the routes inside. In order to support multiple (sub)domains, there could be multiple `routes` folders like `routes_registration.`, `routes_[tenant].` and `routes_admin.[tenant].`. Notice the dot at the end, which means that it is a subdomain. Without the dot, we could support thinks like `site-with-german-title.de` and `english-site.co.uk` as then, it would match against the whole host.

The only thing I don't like about this, however, is the fact that we're breaking the convention, that each folder name equals one segment of the path exactly (as we add the `routes_` prefix.
One way to work around that would be to have a new folder called `domains` in the `src` directory, and inside there other folders with the names / wildcards of the subdomains. However, we'd then need a way to split that part from where the actual path-based routing begins. This could be done with a convention (like a folder named `_routes` in there), however, that would prevent people from having a `_routes` subdomain.

Please let me know what you think about that, I'm happy to hear other opinions!

### Alternatives considered

What I've tried so far is using a reverse proxy (traefik in this case) to map the subdomains to path prefixes (e.g. `registration.my-platform.tld` to `my-platform.tld/registration` internally. This, however, does not work, as the base path is always wrong in that case.
Maybe this is related to #595, but I don't think its the same case as with that proxy it would need a multiple different base paths (or rather hosts).

### Importance

would make my life easier

### Additional Information

This feature is rather important for my application, as there is the requirement for nice urls using subdomains.

I'm also happy to try implementing something once we have a proposed solution and someone can point me in the right direction.

**Comment by Conduitry** [comment]
If each subdomain's application has the same basic shape (which it should if you want to use a single application for all of them), then you should be able to just expose the `Host` header to the rest of your app either in your `handle` hook or in a top-level `+layout.server.js`.

**Comment by hrueger** [comment]
Hi @Conduitry,
thanks for the fast response. I haven't thought about that. However, I don't think that it would work for the described usecase.
The routes are not the same for all subdomains. There's the difference between user-frontend and tenant-admin-backend as well as the registration pages.
They do share a lot of UI components and server logic though, so I'd prefer not to move that into a package make separate apps...

**Comment by hrueger** [comment]
It seems possible with Nuxt: https://ynechaev.medium.com/subdomains-with-nuxt-js-483df826c788

**Comment by georgeiliadis91** [comment]
I would +1 this, I am looking into making an application where each user has his own subdomain and having a build in way to get it would be really convenient. This is something that is supported in other frameworks as far as I remember.

**Comment by mquandalle** [comment]
I have the same use case : ~4 apps that share most of their UI components, stores, utils, but with different routes. The apps are deployed on different domains. ([ex1](https://bonforfait.fr/forfait-mobile), [ex2](https://tariffamobile.it/))

The way I solve it currently is to have a monorepo with a `routes` directory per app, like `routes/app1`. and `routes/app2` I then use the `kit.files.routes` config to set the correct directory depending on an env variable `PUBLIC_APP=app1 npm run dev`.

I deploy these apps on Vercel where I have 4 different "projects" connected to the same repo, but with a different env variable.

This works very well, except that now the deployment is slow. Vercel needs to rebuild the app 4 times and it isn't done in parallel (extra workers are costly on Vercel). If Svelte Kit supported this "multi-tenant application" pattern to build a single "app" to serve my 4 distinct sites, I would be happy to switch to it to simplify the development and deployment processes.

**Comment by hrueger** [comment]
> Is there any reason you need those to instead be a single app?

I'd say it's about convenience for the developer. Svelte Kit is already so great at dev experience and I (and if I understand @mquandalle and @georgeiliadis91 correctly them, too) feel like being forced to split apps - just because the routing technique we'd like to use is partly based on (sub)domains and not exclusively on paths - impairs that quite a bit.

Advantages I see with a single app are the following:
- easier deployment if done "by hand" (not on vercel, etc) as you'd only have a single application to start and not 3 + load balancer which does the routing
- Shorter build time and less complicated build pipeline requirements (see comment by @mquandalle)
- no need for custom logic like parsing the host header and accessing that in the template differently to path parameters
- no need for a monorepo, which can be complicated to setup, especially with TypeScript typings
- benefits of a single app backend instead of multiple (only one db connection, ...)

Don't get me wrong. I could probably achieve the same thing with the current version of Svelte Kit, it would just not be nice code and a good dev experience. You could also completely ignore the router, parse the path segment of the URL and have a single template with a ton of `{#if`s. It would just not be as convenient, usable and (most importantly) maintainable then using the router.
I see the (sub)domain routing feature kind of related to this example.

Can you see my points?

Best regards

**Comment by antony** [comment]
Couldn't you just deploy a single app, and assign all of your desired subdomains to it in vercel's config?

then you can look at the host header in a hook, and just respond accordingly.

single app, single deployment, multiple assigned domains

**Comment by mquandalle** [comment]
> Couldn't you just deploy a single app, and assign all of your desired subdomains to it in vercel's config?

I would like to, but I don't know how to benefit from file system routing for multiple apps simultaneously.

```
routes/
  app1.com/
  app2.fr/
  subdomain.app1.com/
``` 

Maybe it's possible to rewrite the query in a hook to append the host name to the query path so that `http://app1.com/about` is routed into `/app1.com/about/`?

**Comment by Rich-Harris** [comment]
Honestly, this feels like 'scenario solving' to me. Bottom line, these are separate apps and should be treated as such. Introducing significant extra complexity to the routing system just to cut down on build time a bit is entirely the wrong trade-off. 

'I don't want to set up a monorepo' simply isn't a good enough reason to make the framework itself more complicated. This is what monorepos are _for_.

There are workable approaches to do the things described in this thread, so I'm going to close this issue.

**Comment by hrueger** [comment]
OK, if you say that's out of scope, we need to accept that.

@mquandalle Did you manage to get your approach (multiple `routes` folders and switching them with an env variable) working with the VSCode Intellisense? I've just tried it and ran into a lot of problems. When having a single `.svelte-kit`
 directory, multiple processes try to write to the same file (I need to have all apps running when developing).
When having different `.svelte-kit` folders, the Intellisense doesn't like that, as the `tsconfig.json` at the root needs to point to one of the generated `tsconfig.json`s explicitely...

**Comment by hrueger** [comment]
@Rich-Harris is there an example of a monorepo available? I'm struggeling with some points. For example: a `.env` file at the workspace root is not parsed for `$env/static/private`. Copying the `.env` file to all the child packages seems kinda redundant...

**Comment by antony** [comment]
Hi @hrueger - please use discord (https://svelte.dev/chat) for usage questions, where people will be better able to discuss and help with your specific requirements.

**Comment by MentalGear** [comment]
Discord is fun, but it's hard to find things on it and it's closed (you need a login). I always find it much better if there's a proper documentation on publicly available sites like github for issues like these that many people might want to google for.

**Comment by cleytoncarvalho** [comment]
I think that is possible now with Universal Hook Reroute:

https://kit.svelte.dev/docs/hooks#universal-hooks-reroute

**Comment by ravi-ojha** [comment]
While the debate is settled, I was exploring this as I had a small personal use.

@hrueger How about handling subdomain via hooks.

**My usecase was very specific and persona. This code also handles just that. It's not generic and I do not recommend you use this for production systems without due diligence.**

```
// src/hooks.ts
import { redirect } from '@sveltejs/kit';
import type { Handle } from '@sveltejs/kit';

export const handle: Handle = async ({ event, resolve }) => {
    const host = event.request.headers.get('host');

    // Check if the request is from a subdomain in both localhost and production
    if (host) {
        // Assuming production domain is 'example.com'
        if ((host !== 'localhost:3000' && host.endsWith('localhost')) ||
            (host !== 'example.com' && host.endsWith('support.example.com'))) {
            // Redirect to '/help' when accessed via any subdomain
            throw redirect(307, '/help');
        }
    }

    // Continue with normal flow if no subdomain
    return resolve(event);
};
```

**Comment by zicklag** [comment]
This is super easy to do with the reroute hook:

- https://ramonmalcolm.com/articles/dynamic-domain-subdomain-routing-sveltekit-guide/
- https://kit.svelte.dev/docs/hooks#universal-hooks-reroute

I'm planning on using it for the "each user has their own subdomain" use-case.

**Comment by zicklag** [comment]
You can see how I do it in my app here if you want: https://github.com/muni-town/weird/blob/main/src/hooks.ts

You don't need to access the database in that hook. What you do is you check if it's a subdomain, and then you route to a subdomain specific route in your app.

Then when it gets to that route, your route checks the database to see whether or not that subdomain exists in the database or should return a "page not found" because it's not there.

**Comment by zicklag** [comment]
I actually have custom domains working, too!

Basically you could do something like:

- In `hooks.ts` we check if the `url.host == env.PUBLIC_DOMAIN`, if so, we just return the route unchanged.
- Otherwise, we check if it's a subdomain, and we reroute to `\`/subsite/${subdomain}\``
- If it's not a subdomain, we treat it as a custom user domain and we reroute to `\`/subsite/${url.host}\``

Then, in our `/subsite/[usernameOrCustomDomain]` route we check the database to see if `usernameOrCustomDomain` is a username in our database. If it is, we render a subdomain site from the user's data.

If `usernameOrCustomDomain` isn't a username, we look in our database for a registered custom domain, and we render the user's site that has that custom domain set.

---

I also some stuff in there, when the user is configuring their custom domain, it does a little DNS "challenge" to make sure that the domain they are inputting is actually resolving to to the app server, but that's just around how you set the domain, not how you do that routing.

There's also some stuff in there to generate a Traefik reverse proxy configuration for each custom domain, so that we can provide automatic HTTPS certificates for everything. But again, that's kind of unrelated to the routing.

**Comment by RonenEizen** [comment]
Another solution could be to have a separation at the configuration level. You can create a separate config file for each sub-app and specify parameters such as `files`, `appDir`, etc in there. Then you create a separate deployment/development scripts for each app.

**Comment by RonenEizen** [comment]
> Another solution could be to have a separation at the configuration level. You can create a separate config file for each sub-app and specify parameters such as `files`, `appDir`, etc in there. Then you create a separate deployment/development scripts for each app.

Apparently separate config file are not supported. The approach is still valid though by using command args and detecting them within the config file as it is shown in this example: [https://github.com/sveltejs/kit/issues/2973#issuecomment-986050439](https://github.com/sveltejs/kit/issues/2973#issuecomment-986050439)
